rust/src/detect_parser/byte_math.rs

```diff
@@ -0,0 +1,1122 @@
+
+// Fixed position parameter count: bytes, offset, oper, rvalue, result
+// result is not parsed with the fixed position parameters as it's
+// often swapped with optional parameters
+pub const DETECT_BYTEMATH_FIXED_PARAM_COUNT: usize = 5;
+// Optional parameters: endian, relative, string, dce, bitmask
+pub const DETECT_BYTEMATH_MAX_PARAM_COUNT: usize = 10;
+
+#[derive(Debug)]
+enum ResultValue {
+    Numeric(u64),
+    String(String),
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct DetectByteMathData {
+    rvalue_str: *const c_char,
+    result: *const c_char,
+    rvalue: u32,
+    offset: i32,
+    bitmask_val: u32,
+    bitmask_shift_count: u16,
+    id: u16,
+    flags: u8,
+    local_id: u8,
+    nbytes: u8,
+    oper: ByteMathOperator,
+    endian: ByteMathEndian, // big, little, dce
+    base: u8,               // From string or dce
+}
+
+impl Drop for DetectByteMathData {
+    fn drop(&mut self) {
+        unsafe {
+            if !self.result.is_null() {
+                let _ = CString::from_raw(self.result as *mut c_char);
+            }
+            if !self.rvalue_str.is_null() {
+                let _ = CString::from_raw(self.rvalue_str as *mut c_char);
+            }
+        }
+    }
+}
+
+impl Default for DetectByteMathData {
+    fn default() -> Self {
+        DetectByteMathData {
+            local_id: 0,
+            flags: 0,
+            nbytes: 0,
+            offset: 0,
+            oper: ByteMathOperator::None,
+            rvalue_str: std::ptr::null_mut(),
+            rvalue: 0,
+            result: std::ptr::null_mut(),
+            endian: DETECT_BYTEMATH_ENDIAN_DEFAULT,
+            base: DETECT_BYTEMATH_BASE_DEFAULT,
+            bitmask_val: 0,
+            bitmask_shift_count: 0,
+            id: 0,
+        }
+    }
+}
+
+impl DetectByteMathData {
+    pub fn new() -> Self {
+        Self {
+            ..Default::default()
+        }
+    }
+}
+
+fn get_string_value(value: &str) -> Result<u8, ()> {
+    let res = match value {
+        "hex" => DETECT_BYTEMATH_BASE_HEX,
+        "oct" => DETECT_BYTEMATH_BASE_OCT,
+        "dec" => DETECT_BYTEMATH_BASE_DEC,
+        _ => return Err(()),
+    };
+
+    Ok(res)
+}
+
+fn get_oper_value(value: &str) -> Result<ByteMathOperator, ()> {
+    let res = match value {
+        "+" => ByteMathOperator::Addition,
+        "-" => ByteMathOperator::Subtraction,
+        "/" => ByteMathOperator::Division,
+        "*" => ByteMathOperator::Multiplication,
+        "<<" => ByteMathOperator::LeftShift,
+        ">>" => ByteMathOperator::RightShift,
+        _ => return Err(()),
+    };
+
+    Ok(res)
+}
+
+fn get_endian_value(value: &str) -> Result<ByteMathEndian, ()> {
+    let res = match value {
+        "big" => ByteMathEndian::BigEndian,
+        "little" => ByteMathEndian::LittleEndian,
+        "dce" => ByteMathEndian::DCE,
+        _ => return Err(()),
+    };
+
+    Ok(res)
+}
+
+// Parsed as a u64 for validation with u32 {min,max}
```

Comment:
Why parse as u64 if we only accept u32 ?
Why not parse as u32 already ?

Reply:
Parsing as a `u64` and then range checking for u32-min/max avoids treating uint32 max + 1 as a string.

Reply:
Thanks

Do I get it right ?
If suricata parses as u32, and I use `uint32 max + 1` in my signature, it will not fail, because it will fall back to string, and then look for a variable named u32max+1, right ?

Maybe that deserves some more comment (in the caller)